lib/jormungandr/test/migration/migration-test.hs

```diff
@@ -0,0 +1,236 @@
+import Control.Concurrent
+    ( threadDelay )
+import Control.Monad
+    ( forever, void )
+import Control.Retry
+    ( RetryPolicy
+    , constantDelay
+    , limitRetries
+    , limitRetriesByCumulativeDelay
+    , retrying
+    )
+import Control.Tracer
+    ( contramap )
+import Data.Aeson
+    ( Value (..), object, toJSON, (.=) )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import System.Environment
+    ( getArgs )
+import System.Exit
+    ( ExitCode (..), exitFailure, exitWith )
+
+import Control.Lens
+    ( (^.), (^..) )
+import Data.Aeson.Lens
+    ( key, values, _String )
+import Network.Wreq
+    ( get, post, responseBody )
+
+import qualified Data.Text as T
+
+main :: IO ()
+main = do
+    (testAction, launchArgs) <- parseArgs @'Testnet
+
+    cfg <- defaultConfigStdout
+    withTrace cfg "migration-test" $ \tr ->
+        testMain @'Testnet tr 8090 testAction launchArgs >>= exitWith
+
+-- | Something to do while the server is running.
+type TestAction (t :: NetworkDiscriminant) = Trace IO Text -> ApiBase -> IO ExitCode
+
+testMain
+    :: forall t. (DecodeAddress t, EncodeAddress t)
+    => Trace IO Text
+    -> Int
+    -> TestAction t
+    -> [String]
+    -> IO ExitCode
+testMain tr serverPort testAction launchArgs = do
+    let apiBase = mkApiBase serverPort
+    let cmd = Command "cardano-wallet-jormungandr" launchArgs (pure ()) Inherit
+    let tr' = contramap (fmap toText) tr
+    res <- withBackendProcess tr' cmd $ do
+        waitForWalletServer serverPort
+        testAction tr apiBase
+
+    case res of
+        Right st -> pure st
+        Left (ProcessDidNotStart _ e) -> do
+            logError tr ("Failed to start process: " <> T.pack (show e))
+            pure (ExitFailure 13)
+        Left (ProcessHasExited _ ExitSuccess) ->
+            pure ExitSuccess
+        Left (ProcessHasExited _ st) -> do
+            logError tr ("Process exited with status " <> T.pack (show st))
+            pure (ExitFailure 14)
+
+-- | @run@ action
+doRun
+    :: forall t. (DecodeAddress t, EncodeAddress t)
+    => Trace IO Text
+    -> ApiBase
+    -> IO ExitCode
+doRun _ _ = forever $ threadDelay maxBound
+
+testWalletName :: Text
+testWalletName = "test"
+
+-- | @step1@ action. Sets up the database on the old version.
+doStep1
+    :: forall t. (DecodeAddress t, EncodeAddress t)
+    => Trace IO Text
+    -> ApiBase
+    -> IO ExitCode
+doStep1 tr apiBase = do
+    waitForSync apiBase
+    mnem <- mnemonicToText @15 . entropyToMnemonic <$> genEntropy :: IO [Text]
+    let postData = object
+            [ "mnemonic_sentence" .= toJSON mnem
+            , "name" .= testWalletName
+            , "passphrase" .= String "0000000000000000000000000000"
+            ]
+    wal <- post (url apiBase "wallets") (toJSON postData)
+    let walId = wal ^. responseBody . key "id" . _String
+    logInfo tr $ "Create wallet with id " <> walId
+    void $ waitForRestore apiBase walId
```

Comment:
**[question]**

What is the need for this call? It doesn't seem that the state of the wallet would have any influence on the step2 :thinking: ?

Reply:
There is no specific need, but generally, migrations are more likely to fail if the DB tables have some rows in them. That's why it creates a wallet and waits for it to restore.

When we actually implement some migrations, we should add extra test data, but that would depend on what the migrations are.